features/285-design-webhooks/design.md

```diff
@@ -1,129 +1,263 @@
-# next webhooks design
+# Background
 
-## Overview
+Web-hooks would be used to notify external services of events in auth gear.
 
-skygear sends webhooks for events that happen in your app, and are designed to help a developer to monitor events, sync database, and many other use cases.
 
-When one of events is triggered, skygear sends a HTTP POST with a payload to the webhook's configured URL.
+# Proposed Designs
 
-## SYNC and ASYNC hooks
+## Events
+Web-hook events are triggered when some mutating operation is performed by
+auth gear.
 
-There are two types of webhook:
+Each operation will trigger two events: BEFORE and AFTER:
+- BEFORE events would be triggered before the operation is performed; the
+  operation can be aborted by web-hook handler.
+- AFTER events would be trigger after the operation is performed.
 
-- `SYNC` - gear's operation will be blocked until a response is received. Gear should respect `SYNC` hook's response, if it indicates failed, operation should abort and return a formatted error response to the client.
-- `ASYNC` - gear's operation won't be blocked and the response from the webhook is ignored.
+Both events has the same event payload. Web-hook event handlers cannot modify
+the payload.
 
-## Request Headers
 
-Following headers will be added to webhook POST request:
+## Delivery
 
-- `X-Skygear-Webhook-Req-ID`: an ID of the request.
-- `X-Skygear-Webhook-Async`: if is a `async` hook, then the value is `TRUE` otherwise is `FALSE`. 
-- `X-Skygear-Webhook-Signature`: a signature string of the webhook event, a develop should can verify the request by skygear's SDK or manually.
+The events would be delivered to web-hook handlers as an HTTP POST request to
+the handler endpoint.
 
-## Request Payload
+The handler endpoint can be specified as full URL or an absolute path. Absolute
+paths will be resolved to full URL using an inferred URL scheme and authority:
+- from HTTP request URL if the event is generated from it; otherwise,
+- from the tenant configuration.
 
-Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+Each event type can have multiple handlers; the order of deliveries is
+unspecified.
 
-All event payloads follows following payload format:
+BEFORE events will always be delivered before AFTER events. BEFORE events will
+be delivered in well-defined order during a request. AFTER events will be
+delivered in unspecified order.
 
-```json=
+Web-hook handler must be idempotent.
```

Comment:
This one is tricky, what is the reason for it? Is it because we want to relax the constraint of we may call a hook multiple times unexpectedly?

This is a huge trade-off in my opinion, as it impose a constraint of how user write their code (e.g. god know what kind of DB operation they might do with their web hook), and I think it deserves a detailed explanation for us to decided if this trade-off is justified.

Reply:
Writing idempotent handler is a best-practice in web-hook design. If the handler is not idempotent, we cannot retry the web-hook (otherwise unexpected things happen), so the web-hook delivery would be much more unreliable than expected.

Reply:
Best practices vs imposing a constraint to the system are two things.

If we impose a constraint, I expect there is some reason we have to do it; And we have to figure out how we make sure no one makes the mistake / aware of it when they write their own web-hook.

if it is just the best practices, that's just something we would mention in docs somewhere, and maybe warn some features shouldn't be used if they don't do it.

Specifically:

- we don't retry before hook from the spec? So before hook doesn't need to be idempotent if retry is the only reason?
- after hook would be retried automatically so it needs to be idempotent, but we have no way to ensure it is.

In that case it seems more appropriate to just write in spec: "Async after hook is expected to be idempotent"

Reply:
Sorry but I don't see any difference between 'must' and 'expected to be'?

For before events, it would be delivered at most once, so technically there is no need for idempotency. However, it is recommended that no side effect to be performed in it, otherwise it'll need to consider what to do when the operation failed.

Reply:
must is stronger, it meant obligations.

Expected / recommended / best practices etc, are way weaker.

Reply:
I agree 'recommended' and 'best practice' are weaker words. However, in my opinion, both 'must' and 'expected to' means the same: if this requirement is not fulfilled, the system will not work as intended - which is what I want to indicate here.

Reply:
okay I think we are sync here, but just the wording could be misleading. Please make sure the spec is read in a way we want their hook is idempotent, not that it must be idempotent.

Reply:
Still, I'm not seeing how using 'must' is misleading: 'must' is a well-known word used in spec (see RFCs). Would you consider 'should' a better word?

Reply:
I don't think it's better to use 'expect to'. I found 'must' much more clearer:
1. the web-hook system is designed with idempotent handlers in mind.
2. if developer don't write idempotent handlers, the system will not work as we indented.
3. therefore, developer must write idempotent handlers.

Also, the point 2 you mentioned ('the webhook won't works if it is not XXX'), is in fact true; the webhook won't work properly if it's not idempotent.

Reply:
Logically it is not a must -- even if the hook is not idempotent, but if the users don't care about the consequence, it doesn't matter then.

On the other hand, the next sentence in the same paragraph "Web-hook handler must return a status code within the 2XX range.", that must is trivial and easy to understand, because if it doesn't return a 2XX status code, we treat it as failure.

Reply:
I'll compromise and say:
`Web-hook handler should be idempotent, since the same event may be delivered multiple times due to retries.`
Does that sounds okay to you?

Reply:
Sounds fine. thx